METRON-508 Expand Elasticsearch templates to support the standard bro logs

JonZeolla opened the pull request:
## Contributor Comments
This PR makes it easier for someone with an existing bro install to send some of their log files into Metron, based off of a combination of the [bro documentation](https://www.bro.org/sphinx/script-reference/log-files.html) and a fresh install of bro 2.5.  There are future plans to expand on this via [METRON-518](https://issues.apache.org/jira/browse/METRON-518) and [METRON-908](https://issues.apache.org/jira/browse/METRON-908).  Specifically, this attempts to provide initial support the default-on fields of the following logs:
 - [Conn](https://www.bro.org/sphinx/scripts/base/protocols/conn/main.bro.html#type-Conn::Info)
 - [DPD](https://www.bro.org/sphinx-git/scripts/base/frameworks/dpd/main.bro.html#type-DPD::Info)
 - [FTP](https://www.bro.org/sphinx/scripts/base/protocols/ftp/info.bro.html#type-FTP::Info)
 - [Files](https://www.bro.org/sphinx/scripts/base/frameworks/files/main.bro.html#type-Files::Info)
 - [CertsInfo](https://www.bro.org/sphinx/scripts/policy/protocols/ssl/known-certs.bro.html#type-Known::CertsInfo)
 - [SMTP](https://www.bro.org/sphinx/scripts/base/protocols/smtp/main.bro.html#type-SMTP::Info)
 - [SSL](https://www.bro.org/sphinx/scripts/base/protocols/ssl/main.bro.html#type-SSL::Info)
 - [Weird](https://www.bro.org/sphinx/scripts/base/frameworks/notice/weird.bro.html#type-Weird::Info)
 - [Notice](https://www.bro.org/sphinx/scripts/base/frameworks/notice/main.bro.html#type-Notice::Info)
 - [DHCP](https://www.bro.org/sphinx/scripts/base/protocols/dhcp/main.bro.html#type-DHCP::Info)
 - [SSH](https://www.bro.org/sphinx/scripts/base/protocols/ssh/main.bro.html#type-SSH::Info)
 - [Software](https://www.bro.org/sphinx/scripts/base/frameworks/software/main.bro.html#type-Software::Info)
 - [Radius](https://www.bro.org/sphinx/scripts/base/protocols/radius/main.bro.html#type-RADIUS::Info)
 - [X509](https://www.bro.org/sphinx/scripts/base/files/x509/main.bro.html#type-X509::Info)
 - [DevicesInfo](https://www.bro.org/sphinx/scripts/policy/misc/known-devices.bro.html#type-Known::DevicesInfo)


## Testing
1.  Create a working directory and pull in this PR
    ```
    mkdir ~/metron-508
    git clone https://github.com/apache/metron ~/metron-508/metron
    cd ~/metron-508/metron
    git remote add jonzeolla https://github.com/jonzeolla/metron
    git pull jonzeolla METRON-508
    ```
1.  Modify [this](https://github.com/JonZeolla/metron/blob/METRON-508/metron-deployment/vagrant/full-dev-platform/Vagrantfile#L20) to remove `sensors,` (to spin up the real sensors).
    ```
    sed -i '' "s/ansibleSkipTags=.*/ansibleSkipTags=\'quick_dev\'/" metron-deployment/vagrant/full-dev-platform/Vagrantfile
    ```
1.  Start up full-dev.
    ```
    cd metron-deployment/vagrant/full-dev-platform
    vagrant up
    ```
1.  Set up the environment in full-dev.
    ```
    vagrant ssh
    sudo su -
    export PATH=$PATH:/usr/local/bro/bin
    service monit stop && service sensor-stubs stop bro && broctl stop
    yum -y install jq wireshark
    ```
1.  Configure kafka in local.bro.
    ```
    sed -i 's/redef Kafka::logs_to_send = .*/redef Kafka::logs_to_send = set(HTTP::LOG, DNS::LOG, Conn::LOG, DPD::LOG, DHCP::LOG, FTP::LOG, SSH::LOG, SSL::LOG, SMTP::LOG, RADIUS::LOG, Weird::LOG, Files::LOG, Notice::LOG, Software::LOG, Known::CERTS_LOG, Known::DEVICES_LOG, X509::LOG);/' /usr/local/bro/share/bro/site/local.bro
    echo "redef Kafka::debug = \"all\";" >> /usr/local/bro/share/bro/site/local.bro
    echo "redef Known::cert_tracking = ALL_HOSTS;" >> /usr/local/bro/share/bro/site/local.bro
    echo "redef Software::asset_tracking = ALL_HOSTS;" >> /usr/local/bro/share/bro/site/local.bro
    sed -i '86 a @load policy/protocols/dhcp/known-devices-and-hostnames.bro' /usr/local/bro/share/bro/site/local.bro
    ```
1.  Monitor the bro kafka topic
    ```
    # Open a new terminal
    cd ~/metron-508/metron/metron-deployment/vagrant/full-dev-platform
    vagrant ssh
    sudo su -
    export PATH=$PATH:/usr/local/bro/bin:/usr/hdp/current/kafka-broker/bin
    kafka-console-consumer.sh --zookeeper localhost:2181 --topic bro
    ```
1.  Monitor the storm logs.
    ```
    # Open a new terminal
    cd ~/metron-508/metron/metron-deployment/vagrant/full-dev-platform
    vagrant ssh
    sudo su -
    export PATH=$PATH:/usr/local/bro/bin:/usr/hdp/current/kafka-broker/bin
    # Look at the storm logs (The "failed to parse" errors for ip_src_addr and ip_dst_addr are expected, and should be addressed as a part of METRON-939)
    tail -f /var/log/storm/workers-artifacts/indexing-*/*/worker.log | grep -i "org.elasticsearch.index.mapper.MapperParsingException: failed to parse"
    # You may want to evaluate worker.log for other errors, but the prior command is helpful to cut through some of the failed indexing of IPv6 addresses
    ```
1.  Run bro against some public pcaps.
    ```
    # In the first of your three terminals
    # These are kept separate so that the flat file log output won't stomp the prior ones, for ingest validation
    mkdir -p ~/brotmp/nitroba ~/brotmp/example-traffic ~/brotmp/ssh ~/brotmp/ftp ~/brotmp/radius
    wget https://www.bro.org/static/traces/exercise-traffic.pcap -O ~/brotmp/example-traffic/exercise-traffic.pcap
    wget http://downloads.digitalcorpora.org/corpora/network-packet-dumps/2008-nitroba/nitroba.pcap -O ~/brotmp/nitroba/nitroba.pcap
    wget https://www.bro.org/static/traces/ssh.pcap -O ~/brotmp/ssh/ssh.pcap
    wget https://github.com/markofu/pcaps/blob/master/PracticalPacketAnalysis/ppa-capture-files/ftp.pcap?raw=true -O ~/brotmp/ftp/ftp.pcap
    wget https://github.com/EmpowerSecurityAcademy/wireshark/blob/master/radius_localhost.pcapng?raw=true -O ~/brotmp/radius/radius_localhost.pcapng
    cd ~/brotmp/example-traffic
    bro -r exercise-traffic.pcap /usr/local/bro/share/bro/site/local.bro -C
    cd ~/brotmp/nitroba
    bro -r nitroba.pcap /usr/local/bro/share/bro/site/local.bro -C
    cd ~/brotmp/ssh
    bro -r ssh.pcap /usr/local/bro/share/bro/site/local.bro -C
    cd ~/brotmp/ftp
    bro -r ftp.pcap /usr/local/bro/share/bro/site/local.bro -C
    cd ~/brotmp/radius
    editcap -F libpcap radius_localhost.pcapng radius_localhost.pcap
    bro -r radius_localhost.pcap /usr/local/bro/share/bro/site/local.bro -C
    ```
1.  Validate that terminals 2 and 3 don't have any errors that you don't expect.
1.  Verify proper indexing in ES and availability in kibana.
    ```
    # Check around and make sure things look okay
    declare -a exists notexists; for protocol in http dns conn dpd dhcp ftp ssh ssl smtp radius weird files notice software known_certs x509 known_devices; do if [[ $(curl -s -XGET "node1:9200/bro*/_search?q=protocol:${protocol}" | jq '.hits.hits') == '[]' ]]; then notexists+=" ${protocol}"; else exists+=" ${protocol}"; fi; done; if [ ${#notexists[@]} -ne 0 ]; then echo -e "\n\n\033[0mThe following do exist in ES: ${exists[@]}\n\033[0;31mThe following do NOT exist in ES: ${notexists[@]}\033[0m"; else echo 'All of the log types are in ES!  Success!'; fi; unset exists notexists
    # Check Kibana.  For example:  http://node1:5000/app/kibana#/visualize/create?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:now-15y,mode:quick,to:now))&_a=(filters:!(),linked:!f,query:(query_string:(analyze_wildcard:!t,query:'*')),uiState:(),vis:(aggs:!((id:'3',params:(field:protocol,orderBy:'2',size:20),schema:segment,type:terms),(id:'2',schema:metric,type:count)),type:histogram))&indexPattern=bro*&type=histogram
    # OPTIONAL testing
    # Run `/usr/share/elasticsearch/bin/plugin install mobz/elasticsearch-head` and look around
    curl -XGET node1:9200/_cat/indices # First column should be all green
    curl -XGET "node1:9200/bro*/_count" # Check the count of entries in the bro index, you can re-run bro against specific PCAPs and watch this increase, etc.
    ```


## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [X] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?


### For code changes:
- [X] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [X] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

nickwallen commented:
Still need to review, but this is a great addition Jon.  

What pairs nicely with this is to expand the unit tests for the Bro parser to test the other log types that you include in this PR.  Right now only HTTP and DNS are tested there.  We can tackle separately if you like.

JonZeolla commented:
Didn't I do that?

simonellistonball commented:
Looks like there are some good unit tests, though maybe not for every new type you mention, or that could come in. Seems like decent coverage though. I wonder, is it worth adding something in the integration tests as well, especially prior to 777 to avoid accidentally regressing this work when that lands?

nickwallen commented:
Actually, I meant the integration tests.  The stuff under `metron-platform/metron-integration-test/src/main/sample/data/bro`.

simonellistonball commented:
Fantastic! Good to see more integration tests in there and great to see additional data ingested. I'm +1 (non-binding) on this.

JonZeolla commented:
I updated the instructions to reflect the repo name change; hopefully it should work but I won't have a chance to test it out for a couple of days.

JonZeolla commented:
Per @simonellistonball 's comments in #531 I added initial support for the native way that Bro handles tracking DHCP's Client ID field and updated the above instructions appropriately.

JonZeolla commented:
Bump

justinleet commented on metron-platform/metron-parsers/src/test/java/org/apache/metron/parsers/bro/BasicBroParserTest.java:
```diff
@@ -68,27 +68,29 @@ public void testDecimalFormatAssumptions() {
 
 	public void testUnwrappedBroMessage() throws ParseException {
-        String rawMessage = "{\"timestamp\":1449511228.474,\"uid\":\"CFgSLp4HgsGqXnNjZi\",\"source_ip\":\"104.130.172.191\",\"source_port\":33893,\"dest_ip\":\"69.20.0.164\",\"dest_port\":53,\"proto\":\"udp\",\"trans_id\":3514,\"rcode\":3,\"rcode_name\":\"NXDOMAIN\",\"AA\":false,\"TC\":false,\"RD\":false,\"RA\":false,\"Z\":0,\"rejected\":false,\"sensor\":\"cloudbro\",\"type\":\"dns\"}";
+        String rawMessage = "{\"ts\":1449511228.474,\"uid\":\"CFgSLp4HgsGqXnNjZi\",\"id.orig_h\":\"104.130.172.191\",\"id.orig_p\":33893,\"id.resp_h\":\"69.20.0.164\",\"id.resp_p\":53,\"proto\":\"udp\",\"trans_id\":3514,\"rcode\":3,\"rcode_name\":\"NXDOMAIN\",\"AA\":false,\"TC\":false,\"RD\":false,\"RA\":false,\"Z\":0,\"rejected\":false,\"sensor\":\"cloudbro\",\"type\":\"dns\"}";
```
I'd like to see these pulled out and using `@Multiline` for readability.  Given that it's not currently in this format, I'm not opposed to just leaving it, but it does make it easier to understand at a glance.  Given that these all get touched anyway right now, it seems like the perfect time to refactor it.

Example:
```
  /**
   * {
   * "ts":1449511228.474,
   * "uid":"CFgSLp4HgsGqXnNjZi",
   * "id.orig_h":"104.130.172.191",
   * "id.orig_p":33893,
   * "id.resp_h":"69.20.0.164",
   * "id.resp_p":53,
   * "proto":"udp",
   * "trans_id":3514,
   * "rcode":3,
   * "rcode_name":"NXDOMAIN",
   * "AA":false,
   * "TC":false,
   * "RD":false,
   * "RA":false,
   * "Z":0,
   * "rejected":false,
   * "sensor":"cloudbro",
   * "type":"dns"
   * }
   */
  @Multiline
  public static String unwrappedBroMessage;
	public void testUnwrappedBroMessage() throws ParseException {
        JSONObject rawJson = (JSONObject)jsonParser.parse(unwrappedBroMessage);
        JSONObject broJson = broParser.parse(unwrappedBroMessage.getBytes()).get(0);

	String expectedBroTimestamp = "1449511228.474";
      	Assert.assertEquals(broJson.get("bro_timestamp"), expectedBroTimestamp);
        String expectedTimestamp = "1449511228474";
	Assert.assertEquals(broJson.get("timestamp").toString(), expectedTimestamp);

	Assert.assertEquals(broJson.get("ip_src_addr").toString(), rawJson.get("id.orig_h").toString());
	Assert.assertEquals(broJson.get("ip_dst_addr").toString(), rawJson.get("id.resp_h").toString());
	Assert.assertEquals(broJson.get("ip_src_port"), rawJson.get("id.orig_p"));
        Assert.assertEquals(broJson.get("ip_dst_port"), rawJson.get("id.resp_p"));
        Assert.assertEquals(broJson.get("uid").toString(), rawJson.get("uid").toString());
        Assert.assertEquals(broJson.get("trans_id").toString(), rawJson.get("trans_id").toString());
        Assert.assertEquals(broJson.get("sensor").toString(), rawJson.get("sensor").toString());
        Assert.assertEquals(broJson.get("type").toString(), rawJson.get("type").toString());
        Assert.assertEquals(broJson.get("rcode").toString(), rawJson.get("rcode").toString());
        Assert.assertEquals(broJson.get("rcode_name").toString(), rawJson.get("rcode_name").toString());

	Assert.assertTrue(broJson.get("original_string").toString().startsWith("DNS"));
    }
```

justinleet commented:
I did leave one comment about refactoring tests a bit, but I'm +1 regardless of if that change happens or not.  Spun it up in full dev and was able to get all the types to show up in the histogram and things looked good all around.

Thanks for the great contribution!

JonZeolla commented:
Thanks.  I haven't used org.adrianwalker.multilinestring.Multiline before, took a quick shot with two of them, I'll do the rest when I have more time.

justinleet commented on metron-platform/metron-parsers/src/test/java/org/apache/metron/parsers/bro/BasicBroParserTest.java:
```diff
@@ -241,8 +380,754 @@ public void testFilesBroMessage() throws ParseException {
+
 	@SuppressWarnings("rawtypes")
 	public void testProtocolKeyCleanedUp() throws ParseException {
 		String rawMessage = "{\"ht*tp\":{\"ts\":1402307733.473,\"uid\":\"CTo78A11g7CYbbOHvj\",\"id.orig_h\":\"192.249.113.37\",\"id.orig_p\":58808,\"id.resp_h\":\"72.163.4.161\",\"id.resp_p\":80,\"trans_depth\":1,\"method\":\"GET\",\"host\":\"www.cisco.com\",\"uri\":\"/\",\"user_agent\":\"curl/7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3\",\"request_bod ...
```
One more multiline

justinleet commented:
@JonZeolla Can you deconflict this?

justinleet commented:
+1 still stands, pending Travis.

nickwallen commented on metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/files/bro_index.template:
```diff
@@ -238,6 +238,538 @@
+          "type": "string",
+          "index": "not_analyzed"
+        },
+        "next_protocol": {
```
Should I expect to see every field in each of the supported log sources defined in the template?

I quickly spot-checked DPD and I don't see `proto`, `disabled_aids` or `packet_segment` defined in the template.  I believe these are all valid fields that a user could get as part of the [DPD record](https://www.bro.org/sphinx-git/scripts/base/frameworks/dpd/main.bro.html#type-DPD::Info).

Let me know if my assumption is wrong.

JonZeolla replied:
Because you can create arbitrary fields in log sources, I decided to only support the default-on fields with this first PR.  

`proto` exists [here](https://github.com/JonZeolla/metron/blob/7e761480c9749d67acfa7de538f54eee96dcba05/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/files/bro_index.template#L215-L218) (i.e. it already existed in another bro log, so in these scenarios I made sure that the type works for both - a more preferred alternative would be to break out individual bro logs into separate indexes), `disabled_aids` is an [internal only field](https://github.com/bro/bro/blob/57da2d091b30aad52d52fce8018feeb2cdf8ff1f/scripts/base/frameworks/dpd/main.bro#L25-L27), and `packet_segment` was explicitly omitted, as it requires that `policy/frameworks/dpd/packet-segment-logging.bro` be loaded.

nickwallen replied:
Ok, cool!  I am sure all these fields were fun to hunt down.

nickwallen commented on metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/files/bro_index.template:
```diff
@@ -238,6 +238,538 @@
+          "type": "string",
+          "index": "not_analyzed"
+        },
+        "failure_reason": {
```
It would be handy to know which fields in the template belong to which Bro log source (DPD vs Conn).  This might make it easier for users (and us) to maintain or customize the template going forward.  I can think of two possible ways to do this, but maybe there are better ways.  

**Option 1:** We can use multiple templates that get [merged into a single final template](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html#multiple-templates).  This would provide a nice, clean separation between the different log sources.  But it might take more effort than it is worth.

**Option 2:**  We can embed `/* c-style block comments */` in the template itself. We could throw a comment above each set of fields that pertain to a given source.  We could also maintain the handy links that you embedded in the PR description.  

```
/* 
 *  DPD
 * https://www.bro.org/sphinx-git/scripts/base/frameworks/dpd/main.bro.html#type-DPD::Info
 */

"failure_reason": {
 ..
```

nickwallen replied:
This suggestion is not necessary for the PR, but I wanted to get your thoughts on this. Consider this extra credit.

JonZeolla replied:
Right, I considered both options, and implemented option 2 at one point, but I removed the comments because of the field name collisions (i.e. two separate bro logs with an overlapping field name).  While reading through the template, it was confusing that a given section wouldn't contain all of the fields for a specific log, because they were addressed earlier in the template under another log's section.

I would prefer to merge this in as-is, and address the collision problem separately (at least, that was my intent).  The first true solution that comes to mind is to put the individual bro logs into distinct indexes, but then we would need to change anywhere in Metron that touches bro data.  I would prefer to do that after METRON-939 (#619), if it gets merged.

nickwallen replied:
The refactor of this template that you suggest seems scary to me without some comments.  Duplicate names seem also like something that should be called out in comments.  There is some useful information here that can only be found in this PR and won't live-on with the code, unfortunately.

But at least it looks like you have one test record per new log type, so the integration tests should fail if a refactor of the index template goes badly.  **I think?**

ottobackwards replied:
I'm sorry if this is left field but..... What if, we provide more than one 'example' template in the bro parser extension, and not try to come up with the perfect configuration?

Also - if and when 777 hits and we move on to 'create an instance of a parser - clone and edit configurations' they can use those samples to do so.

JonZeolla replied:
@nickwallen I think we delay the details of that conversation until/if the change occurs.  I can  comment any fields that are used in multiple logs, as that's probably worthwhile keeping in the template itself.

@ottobackwards Well, I expect people to customize this if they already have a bro install prior to using Metron, as they will probably have custom and non-default fields turned on.  If not, I would want the template to have full support of sending data through a default bro install (the intent of this PR).

JonZeolla commented:
@nickwallen So, I'm not entirely done with the documentation but I pushed it out for a quick, general review.  In doing this, I noticed that some new default-on fields were added with the release of 2.5/2.5.1 (for example, `server_appdata` in [SSL](https://www.bro.org/sphinx/scripts/base/protocols/ssl/main.bro.html#type-SSL::Info) was added with 2.5).  For now, I'm going to ignore those.

JonZeolla commented:
@nickwallen @justinleet Can I get a quick review of my recent changes before I merge?  Mostly focused on how I approached documenting the template, but feel free to comment on anything.

I just spun this up in full-dev and verified that my above steps are still valid and this works as intended.  At this point I'm iteratively crossing my fingers and retrying Travis until it gets the whole way through and gives me a result.  Can't wait for #624

JonZeolla commented:
Last try before I merge in #624 and make it a dependancy.

JonZeolla commented:
Out of sheer frustration I merged in #624 in its current state and got Travis to pass on the first try after a dozen failures.  Great work @justinleet @cestella and others.  Obviously now this has a dependency of METRON-1004 so I've edited the title of this to include "[DO NOT MERGE UNTIL METRON-1004]".  Still looking for a review - looking back I should have squash merged, but due to the simplicity of the changes in this PR, it's probably not worth going back and fixing it.

justinleet commented:
@JonZeolla 1004 is in, but there's been a few changes in, so I'd ask that you double check that things still line up as expected.  I'm going to look back through the changes this morning on here this morning, now that Travis actually works.

JonZeolla commented:
Should be good to go now, pending Travis.

justinleet commented:
+1. I'm good with the most recent changes, pending Travis.

JonZeolla commented:
I spotted a duplicate in taking a scan of the template, so I ran `grep '": {' metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/files/bro_index.template | sort | uniq -c | grep -v 1` to make sure that was the only one.  I fixed it and pushed things up, along with other minor tweaks - Travis was successful so I plan to merge soon.  I also put some thoughts for a follow-on PR in [METRON-1010](https://issues.apache.org/jira/browse/METRON-1010).
